[zi-yii] iP

zi-yii opened the pull request:
# 🤖 BobTheBot  

> "You’re braver than you believe, stronger than you seem and smarter than you think." - Christopher Robin [(source)](https://parade.com/935628/parade/winnie-the-pooh-quotes/)

Bob will lend you a helping hand in keeping track of everything you need to do. It is
- text-based
- **simple** to understand, learn and use
- ~~FAST~~ _SUPER_ FAST to use

All you have to do is:
1. Download it from [here](https://github.com/zi-yii/ip)
2. Start adding your tasks
3. Interact with Bob to manage your tasks 😃  

Functionalities:
- [ ] Add different task types
- [x] Mark or Unmark tasks
- [ ] Delete tasks
- [ ] Find task by keyword

If you are a Java programmer, you can use it to practise Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

***LET BOB BE YOUR BEST BUDDY NOW !!***

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

zi-yii committed:
Update name and greetings of chatbot.

zi-yii committed:
Add echo function

zi-yii committed:
Add user input history

zi-yii committed:
Create Task class

zi-yii committed:
Add ability to mark tasks as done and undone

zi-yii committed:
Add support for tracking three types of tasks

zi-yii committed:
Test using I/O redirection technique

zi-yii committed:
Handle errors

zi-yii committed:
Add delete function

zi-yii committed:
Use Enumerations

zi-yii committed:
Add save task function

zi-yii committed:
Merge branch 'branch-Level-7' into master

zi-yii committed:
Change format for deadline dates

zi-yii committed:
Add ability to obtain relevant tasks for a specific date

zi-yii committed:
Merge branch 'branch-Level-8' into master

zi-yii committed:
Extract TaskList class

zi-yii committed:
Extract Storage class and Parser class

zi-yii committed:
Extract Ui class

zi-yii committed:
Merge branch 'branch-A-MoreOOP' into master

zi-yii committed:
Organise into packages bob and bob.task

zi-yii committed:
Merge branch 'branch-A-Packages' into master

zi-yii committed:
Merge branch 'add-gradle-support' into master

zi-yii committed:
Add ParserTest.java and TaskListTest.java

zi-yii committed:
Merge branch 'branch-A-JUnit' into master

zi-yii committed:
Use Gradle to run JUnit tests

zi-yii committed:
Merge branch 'branch-A-Gradle' into master

zi-yii committed:
Add JavaDoc comments

zi-yii committed:
Tweak code to comply with coding standard

zi-yii committed:
Add ability to find task by keyword

zi-yii committed:
Merge branch 'branch-A-JavaDoc' into master

zi-yii committed:
Merge branch 'branch-A-CodingStandard' into master

Resolution: keep both

zi-yii committed:
Merge branch 'branch-Level-9' into master

Resolution: keep both

zi-yii committed:
Update expected error message

Change expected error message for
delTask_invalidTaskNum_exceptionThrown() to
"The task number provided is invalid."

KennyLewi commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,355 @@
+
+                case LIST:
+                    commandList();
+                    break;
+
+                case RELEVANT:
+                    commandRelevant(taskDetails);
+                    break;
+
+                case MARK:
+                    commandMark(taskDetails);
+                    break;
+
+                case UNMARK:
+                    commandUnmark(taskDetails);
+                    break;
+
+                case TODO:
+                    commandTodo(taskDetails);
+                    break;
+
+                case DEADLINE:
+                    commandDeadline(taskDetails);
+                    break;
+
+                case EVENT:
+                    commandEvent(taskDetails);
+                    break;
+
+                case DELETE:
+                    commandDelete(taskDetails);
+                    break;
+
+                case FIND:
+                    commandFind(taskDetails);
+                    break;
+
+                case UNKNOWN:
+                    // Fallthrough
+
```
Maybe you can remove the empty lines in between the cases?

KennyLewi commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,183 @@
+
+     * @return true if the task list is empty, false otherwise.
+     */
+    boolean isEmpty() {
```
maybe add the public keyword?

KennyLewi commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,183 @@
+     * @return A List of Task objects.
+     */
+
+    List<Task> getTasks() {
```
maybe add the public keyword? (same for the other methods in this class)

KennyLewi commented on src/main/java/bob/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @return 1 if task is completed, 0 otherwise.
+     */
+    int isDoneBinary() {
```
maybe add the public keyword?

KennyLewi commented on src/test/java/bob/ParserTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    public void testParseDateTime_validFormat() {
+        String validFormatStr = "2024-12-12 1212";
+        LocalDateTime expectedDateTime = LocalDateTime.of(2024, 12, 12, 12, 12);
+        try {
+            assertEquals(expectedDateTime, Parser.parseDateTime(validFormatStr));
+        } catch (BobException e) {
+            fail("BobException should not be thrown for a valid date time format.");
+        }
+    }
+
+
+    @Test
+    public void parseDateTime_invalidTime_exceptionThrown() {
+        String invalidTimeStr = "2024-12-12 99:99";
+        BobException e = assertThrows(BobException.class, () -> Parser.parseDateTime(invalidTimeStr));
+        assertEquals("Please provide the correct date and 24-hour time format: "
+                + "yyyy-mm-dd HHmm"
+                + "\nEg. 2024-08-27 1530 for Aug 27 2024 03.30pm", e.getMessage());
+    }
+
+    @Test
+    public void parseDateTime_invalidDate_exceptionThrown() {
+        String invalidDateStr = "2024-31-00 1600";
+        BobException e = assertThrows(BobException.class, () -> Parser.parseDateTime(invalidDateStr));
+        assertEquals("Please provide the correct date and 24-hour time format: "
+                + "yyyy-mm-dd HHmm"
+                + "\nEg. 2024-08-27 1530 for Aug 27 2024 03.30pm", e.getMessage());
+    }
+
+    @Test
+    public void parseDateTime_invalidFormat_exceptionThrown() {
+        String invalidFormatStr = "Dec 12 2024 4pm";
+        BobException e = assertThrows(BobException.class, () -> Parser.parseDateTime(invalidFormatStr));
+        assertEquals("Please provide the correct date and 24-hour time format: "
+                + "yyyy-mm-dd HHmm"
+                + "\nEg. 2024-08-27 1530 for Aug 27 2024 03.30pm", e.getMessage());
+    }
```
maybe you can add javadoc comments for your testcases for better understanding?

KennyLewi commented on src/test/java/bob/TaskListTest.java:
```diff
@@ -0,0 +1,113 @@
+    @Test
+    void testGetRelevantTasks_validDate() throws BobException {
+        String validDate1 = "2024-12-12";
+        String expected1 = "1. [D][ ] submit assignment (by: Dec 12 2024 11:59 pm)\n"
+                + "2. [E][ ] trip (from: Dec 09 2024 12:00 am to: Dec 16 2024 12:00 am)\n"
+                + "Total number of relevant tasks for Dec 12 2024: 2";
+        assertEquals(expected1, taskList.printRelevantTasksByDate(validDate1));
+
+        String validDate2 = "2024-12-09";
+        String expected2 = "1. [E][ ] trip (from: Dec 09 2024 12:00 am to: Dec 16 2024 12:00 am)\n"
+                + "Total number of relevant tasks for Dec 09 2024: 1";
+        assertEquals(expected2, taskList.printRelevantTasksByDate(validDate2));
+
+        String validDate3 = "2024-12-10";
+        String expected3 = "1. [E][ ] project meeting (from: Dec 10 2024 02:00 pm to: Dec 10 2024 04:00 pm)\n"
+                + "2. [E][ ] trip (from: Dec 09 2024 12:00 am to: Dec 16 2024 12:00 am)\n"
+                + "Total number of relevant tasks for Dec 10 2024: 2";
+        assertEquals(expected3, taskList.printRelevantTasksByDate(validDate3));
+    }
+
+    @Test
+    void testGetRelevantTasks_noRelevantTasks() throws BobException {
+        String irrelevantDate = "2025-01-01";
+        String expected = "Total number of relevant tasks for Jan 01 2025: 0";
+        assertEquals(expected, taskList.printRelevantTasksByDate(irrelevantDate));
+    }
+
+    @Test
+    void getRelevantTasks_invalidDateFormat_exceptionThrown() {
+        String invalidDate = "01-01-2025";
+        BobException e = assertThrows(BobException.class, () -> taskList.printRelevantTasksByDate(invalidDate));
+        assertEquals("Invalid date format. Required format: relevant yyyy-MM-dd", e.getMessage());
+    }
+
+    @Test
+    void getRelevantTasks_emptyDate_exceptionThrown() {
+        String emptyDateStr = "";
+        BobException e = assertThrows(BobException.class, () -> taskList.printRelevantTasksByDate(emptyDateStr));
+        assertEquals("Invalid date format. Required format: relevant yyyy-MM-dd", e.getMessage());
+    }
```
maybe you can add javadoc comments for your @BeforeEach and each testcase?

KennyLewi commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,355 @@
+     * @param taskDetails The description, start date, and end date of the Event task.
+     * @throws BobException If there are missing details, wrong format, or invalid dates.
+     */
+    static void commandEvent(String taskDetails) throws BobException {
```
maybe you can rename commandEvent to commandAddEvent to make it clearer as a method name?

KennyLewi reviewed (commented):
Overall, LGTM, just a few nits. Another suggestion is maybe you can let the parser handle the command switching, instead of letting Bob.java handle the switch cases? I liked how you added javadoc comments to most of your methods and classes, and how thorough your test cases are

zi-yii committed:
Use checkstyle to detect coding style violations

zi-yii committed:
Merge branch 'branch-A-Checkstyle' into master

zi-yii committed:
Connect Bob chatbot functionality to GUI

zi-yii committed:
Update GUI design layout

zi-yii committed:
Merge branch 'branch-Level-10' into master

zi-yii committed:
Use varargs for showMessage method in Ui class

zi-yii committed:
Merge branch 'branch-A-Varargs' into master

zi-yii committed:
Add assertions

The code lacks validation checks and may lead to undetected bugs, such
as null task list.

This needs to change to prevent bugs.

Add assertions to various classes to help with error handling and
prevent bugs.

Adding assertions allows us to catch potential bugs early during
development and improve code reliability.

zi-yii committed:
Improve code quality

The code in Bob, Parser, and TaskList lacks clarity and
consistency. Some methods handle repetitive tasks inefficiently. Date
parsing logic was scattered across classes, reducing maintainability.

This needs to change to ensure better code organisation and readability.

Move date parsing logic to Parser for centralised handling. Refactor
TaskList methods to eliminate code duplication by extracting methods
(e.g., buildTaskListString). Introduce validateEventDates and
validateTaskDetails methods to encapsulate validation logic,
ensuring better abstraction

These changes are done to adhere to DRY principles, focus on
abstraction, centralise logic for easier updates, and make the codebase
easier to understand and extend.

zi-yii committed:
Use Streams in TaskList class

Loops are used to find matching tasks with the specified date or
keyword.

Loops may not be as easy to understand.

Replace loops with streams to find matching tasks by using the 'filter'
stream operation.

This change enhances readability and performance in task filtering.

zi-yii committed:
Extract setCommandType method

zi-yii committed:
Merge pull request #2 from zi-yii/branch-A-Assertions

Add assertions

zi-yii committed:
Merge branch 'master' of https://github.com/zi-yii/ip

* 'master' of https://github.com/zi-yii/ip:
  Add assertions

zi-yii committed:
Merge branch 'master' into branch-A-CodeQuality

zi-yii committed:
Merge pull request #3 from zi-yii/branch-A-CodeQuality

Improve code quality

zi-yii committed:
Merge branch 'master' of https://github.com/zi-yii/ip

* 'master' of https://github.com/zi-yii/ip:
  Extract setCommandType method
  Improve code quality

zi-yii committed:
Merge branch 'master' into branch-A-Streams

Resolution: keep both

zi-yii committed:
Merge pull request #4 from zi-yii/branch-A-Streams

Use streams in TaskList.java

zi-yii committed:
Merge branch 'master' of https://github.com/zi-yii/ip

* 'master' of https://github.com/zi-yii/ip:
  Use Streams in TaskList class

zi-yii committed:
Add ability to sort tasks by description and by date

zi-yii committed:
Merge branch 'branch-C-Sort' into master

zi-yii committed:
Add title to GUI window

zi-yii committed:
Merge branch 'branch-Level-10' into master

zi-yii committed:
Edit header comment for main method in Launcher class

zi-yii committed:
Merge branch 'branch-A-CodingStandard' into master

zi-yii committed:
Reduce padding between dialog boxes

zi-yii committed:
Merge branch 'branch-Level-10' into master

zi-yii committed:
Add screenshot of GUI window

zi-yii committed:
Merge branch 'branch-ProductScreenshot' into master

zi-yii committed:
Add tests for printTasksbyKeyword method from TaskList

zi-yii committed:
Merge branch 'branch-A-JUnit' into master

zi-yii committed:
Add ability to close the GUI window automatically on command "bye"

zi-yii committed:
Merge branch 'branch-Level-10' into master

zi-yii committed:
Add user guide

zi-yii committed:
Merge branch 'branch-A-UserGuide' into master

zi-yii committed:
Update README file

zi-yii committed:
Merge branch 'branch-A-UserGuide' into master

zi-yii committed:
Update user guide

zi-yii committed:
Merge branch 'branch-A-UserGuide' into master

zi-yii committed:
Edit README files

zi-yii committed:
Merge branch 'branch-A-UserGuide' into master

zi-yii committed:
Remove alerts

zi-yii committed:
Merge branch 'branch-A-UserGuide' into master

zi-yii committed:
Add link for location of jar file

zi-yii committed:
Merge branch 'branch-A-UserGuide' into master
